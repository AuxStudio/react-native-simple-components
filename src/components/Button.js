import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

import Icon from '../assets/icons/index';
import styleConstants from '../assets/styleConstants';

import Touchable from './Touchable';

const styles = StyleSheet.create({
    button: {
        ...styleConstants.smallShadow,
        width: styleConstants.windowWidth - 32,
        height: 56,
        flexDirection: 'row',
        paddingHorizontal: 16,
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderColor: styleConstants.white,
        marginTop: 16,
        alignSelf: 'center',
    },
    disabled: {
        opacity: 0.33,
    },
    icon: {
        position: 'absolute',
        left: 16,
        fontSize: styleConstants.iconFont,
    },
    text: {
        fontSize: styleConstants.regularFont,
    },
});

export default (Button = props => {
    /*
        PROPTYPES

        backgroundColor: PropTypes.string,
        iconName: PropTypes.string,
        disabled: PropTypes.string,
        text: PropTypes.string.isRequired,
        handlePress: PropTypes.func.isRequired,
        androidRipple: PropTypes.bool,
        androidRippleColor: PropTypes.string,
        style: PropTypes.oneOfType([
            PropTypes.object,
            PropTypes.number,
        ]),
    */

    const altColor =
        props.backgroundColor === 'transparent' ||
        props.backgroundColor === styleConstants.primary
            ? styleConstants.white
            : styleConstants.primary;

    const icon = props.iconName ? (
        <Icon
            name={props.iconName}
            style={[styles.icon, { color: altColor }]}
        />
    ) : null;

    const button = props.disabled ? (
        <View
            style={[
                styles.button,
                styles.disabled,
                { backgroundColor: props.backgroundColor },
                props.style,
            ]}>
            {icon}
            <Text
                style={[
                    styles.text,
                    { color: altColor },
                    styleConstants.primaryFont,
                ]}>
                {props.text}
            </Text>
        </View>
    ) : (
        <Touchable
            onPress={props.handlePress}
            style={[
                styles.button,
                { backgroundColor: props.backgroundColor },
                props.style,
            ]}
            androidRipple
            androidRippleColor={altColor}>
            {icon}
            <Text
                style={[
                    styles.text,
                    { color: altColor },
                    styleConstants.primaryFont,
                ]}>
                {props.text}
            </Text>
        </Touchable>
    );

    return <View>{button}</View>;
});
