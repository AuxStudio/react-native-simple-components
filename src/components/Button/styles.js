import { StyleSheet } from 'react-native';

import styleConstants from '../../styleConstants';

const styles = StyleSheet.create({
  button: {
    alignSelf: 'stretch',
    flexDirection: 'row',
    padding: styleConstants.dimensions.padding.large,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: styleConstants.colors.dividerColor,
  },
  textLeftContainer: {
    justifyContent: 'flex-start',
  },
  iconContainer: {
    position: 'absolute',
    left: styleConstants.dimensions.padding.large,
  },
  iconRight: {
    left: 'auto',
    right: styleConstants.dimensions.padding.large,
  },
  textLeftIcon: {
    position: 'relative',
    left: 0,
    marginRight: styleConstants.dimensions.padding.small,
  },
  icon: {
    fontSize: styleConstants.fonts.sizes.icon,
    color: styleConstants.colors.primaryText,
  },
  text: {
    fontSize: styleConstants.fonts.sizes.regular,
    color: styleConstants.colors.primaryText,
  },
});

export default styles;
