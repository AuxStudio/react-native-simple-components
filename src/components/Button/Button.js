import React from 'react';
import { View, Text, ActivityIndicator } from 'react-native';

import styles from './styles';
import styleConstants from '../assets/styleConstants';

import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import Touchable from './Touchable';

export default (Button = (props) => {
  /*
        PROPTYPES

        iconName: PropTypes.string,
        customIcon: PropTypes.node, 
        iconRight: PropTypes.bool,
        text: PropTypes.string.isRequired,
        textLeft: PropTypes.bool, // moves text left
        handlePress: PropTypes.func.isRequired,
        disabled: PropTypes.string,
        showLoader: PropTypes.bool, // will display an ActivityIndicator
        loaderColor: PropTypes.string,

        androidRipple: PropTypes.bool,
        androidRippleColor: PropTypes.string,
        showShadow: PropTypes.bool,
        // style: PropTypes.node,
        // iconStyle: PropTypes.node,
        // textStyle: PropTypes.node,
        // disabledStyle: PropTypes.node,
    */

  const textLeftContainerStyles = props.textLeft && styles.textLeftContainer;
  const textLeftIconStyles = props.textLeft && styles.textLeftIcon;

  const iconRightStyles = props.iconRight && styles.iconRight;

  const icon =
    !props.showLoader && props.customIcon ? (
      props.customIcon
    ) : props.iconName ? (
      <MaterialIcon
        name={props.iconName}
        style={[styles.icon, textLeftIconStyles, iconRightStyles, props.iconStyle]}
      />
    ) : null;

  const shadowStyles = props.showShadow && {
    ...styleConstants.smallShadow,
  };

  const text = !props.showLoader ? (
    <Text style={[styles.text, props.textStyle]}>{props.text}</Text>
  ) : null;

  const loader = props.showLoader && <ActivityIndicator size="small" color={props.loaderColor} />;

  const button = props.disabled ? (
    <View
      style={[
        styles.button,
        props.disabledStyle ? props.disabledStyle : styles.disabled,
        shadowStyles,
        textLeftContainerStyles,
        props.style,
      ]}
    >
      {icon}
      {text}
      {loader}
    </View>
  ) : (
    <Touchable
      onPress={props.handlePress}
      style={[styles.button, shadowStyles, textLeftContainerStyles, props.style]}
      androidRipple={props.androidRipple}
      androidRippleColor={props.androidRippleColor}
    >
      {icon}
      {text}
      {loader}
    </Touchable>
  );

  return button;
});
