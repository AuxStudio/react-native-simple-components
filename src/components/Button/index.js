import React from 'react';
import PropTypes from 'prop-types';
import { View, Text, ActivityIndicator, ViewPropTypes } from 'react-native';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';
import Touchable from '../Touchable';

const propTypes = {
  iconName: PropTypes.string,
  customIcon: PropTypes.node,
  iconRight: PropTypes.bool,
  text: PropTypes.string,
  textLeft: PropTypes.bool, // moves text left
  handlePress: PropTypes.func,
  disabled: PropTypes.bool,
  showLoader: PropTypes.bool, // will display an ActivityIndicator
  loaderColor: PropTypes.string,
  androidRipple: PropTypes.bool,
  androidRippleColor: PropTypes.string,
  androidRippleBorderless: PropTypes.bool,
  showShadow: PropTypes.bool,
  style: ViewPropTypes.style,
  iconContainerStyle: ViewPropTypes.style,
  iconStyle: Text.propTypes.style,
  textStyle: Text.propTypes.style,
  testID: PropTypes.string,
};

const defaultProps = {
  text: 'Button',
};

const Button = ({
  iconName,
  customIcon,
  iconRight,
  text,
  textLeft,
  handlePress,
  disabled,
  showLoader,
  loaderColor,
  androidRipple,
  androidRippleColor,
  androidRippleBorderless,
  showShadow,
  style,
  iconContainerStyle,
  iconStyle,
  textStyle,
  testID,
}) => {
  const textLeftContainerStyles = textLeft && styles.textLeftContainer;
  const textLeftIconStyles = textLeft && !iconRight && styles.textLeftIcon;
  const iconRightStyles = iconRight && styles.iconRight;
  const shadowStyles = showShadow && styleConstants.shadows.small;

  let iconComponent;

  if (!showLoader && customIcon) {
    iconComponent = (
      <View style={[styles.iconContainer, textLeftIconStyles, iconRightStyles, iconContainerStyle]}>
        {customIcon}
      </View>
    );
  } else if (iconName) {
    iconComponent = (
      <View style={[styles.iconContainer, textLeftIconStyles, iconRightStyles, iconContainerStyle]}>
        <MaterialIcon name={iconName} style={[styles.icon, iconStyle]} />
      </View>
    );
  }

  const textComponent = !showLoader ? <Text style={[styles.text, textStyle]}>{text}</Text> : null;

  const loaderComponent = showLoader && <ActivityIndicator size="small" color={loaderColor} />;

  return (
    <Touchable
      onPress={handlePress}
      style={[styles.button, shadowStyles, textLeftContainerStyles, style]}
      androidRipple={androidRipple}
      androidRippleColor={androidRippleColor}
      androidRippleBorderless={androidRippleBorderless}
      disabled={disabled}
      testID={testID}
    >
      {iconComponent}
      {textComponent}
      {loaderComponent}
    </Touchable>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
