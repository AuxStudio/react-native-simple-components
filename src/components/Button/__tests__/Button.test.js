import React from 'react';
import renderer from 'react-test-renderer';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';

import Button from '../Button';

it('renders a Button', () => {
  expect(
    renderer.create(
      <Button
        iconName="check"
        text="Test"
        handlePress={jest.fn()}
        androidRipple
        androidRippleColor="red"
        androidRippleBorderless
        showShadow
      />,
    ),
  ).toMatchSnapshot();
});

it('renders a Button with text left and icon right', () => {
  expect(
    renderer.create(<Button iconName="check" iconRight text="Test" textLeft />),
  ).toMatchSnapshot();
});

it('renders a Button with a customIcon', () => {
  expect(renderer.create(<Button customIcon={<MaterialIcon name="check" />} />)).toMatchSnapshot();
});

it('renders a Button with a loader', () => {
  expect(renderer.create(<Button showLoader loaderColor="red" />)).toMatchSnapshot();
});

it('renders a disabled Button', () => {
  expect(renderer.create(<Button disabled />)).toMatchSnapshot();
});

it('renders a Button with no props', () => {
  expect(renderer.create(<Button />)).toMatchSnapshot();
});
