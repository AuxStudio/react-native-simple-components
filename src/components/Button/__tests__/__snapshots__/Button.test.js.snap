// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Button 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "borderColor": null,
      "borderWidth": 0,
      "elevation": 2,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
      "shadowColor": "#212121",
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.17,
      "shadowRadius": 2,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "left": 16,
          "position": "absolute",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#212121",
              "fontSize": 24,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#212121",
          "fontSize": 17,
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`renders a Button with a customIcon 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "left": 16,
          "position": "absolute",
        },
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          undefined,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#212121",
          "fontSize": 17,
        },
        undefined,
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`renders a Button with a loader 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="red"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`renders a Button with no props 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#212121",
          "fontSize": 17,
        },
        undefined,
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`renders a Button with text left and icon right 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "flexDirection": "row",
      "justifyContent": "flex-start",
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "left": 16,
          "position": "absolute",
        },
        false,
        Object {
          "left": "auto",
          "right": 16,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#212121",
              "fontSize": 24,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#212121",
          "fontSize": 17,
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`renders a disabled Button 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "stretch",
      "backgroundColor": "#E0E0E0",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#212121",
          "fontSize": 17,
        },
        undefined,
      ]
    }
  >
    Button
  </Text>
</View>
`;
